File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/presentation/ui/download/DownloadViewModel.java
package com.bits.bee.bpmc.presentation.ui.download;

import java.lang.System;

/**
 * Created by aldi on 29/03/22.
 */
@dagger.hilt.android.lifecycle.HiltViewModel()
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u00006\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\b\n\u0000\n\u0002\u0010\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0002\b\u001d\b\u0007\u0018\u00002\u000e\u0012\u0004\u0012\u00020\u0002\u0012\u0004\u0012\u00020\u00030\u0001:\u0001,B\u0017\b\u0007\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u00a2\u0006\u0002\u0010\bJ\u0011\u0010\u000b\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u000e\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\b\u0010\u000f\u001a\u00020\u0010H\u0002J\u0011\u0010\u0011\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u0012\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u0013\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u0014\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u0015\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u0016\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u0017\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u0018\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u0019\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u001a\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u001b\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u001c\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u001d\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u001e\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\u001f\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010 \u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010!\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\"\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010#\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010$\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010%\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010&\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010\'\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010(\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010)\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010*\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rJ\u0011\u0010+\u001a\u00020\fH\u0082@\u00f8\u0001\u0000\u00a2\u0006\u0002\u0010\rR\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\t\u001a\u00020\nX\u0082\u000e\u00a2\u0006\u0002\n\u0000\u0082\u0002\u0004\n\u0002\b\u0019\u00a8\u0006-"}, d2 = {"Lcom/bits/bee/bpmc/presentation/ui/download/DownloadViewModel;", "Lcom/bits/bee/bpmc/presentation/base/BaseViewModel;", "Lcom/bits/bee/bpmc/presentation/ui/download/DownloadState;", "Lcom/bits/bee/bpmc/presentation/ui/download/DownloadViewModel$UIEvent;", "districtRepository", "Lcom/bits/bee/bpmc/domain/repository/DistrictRepository;", "di", "Lcom/bits/bee/bpmc/domain/usecase/download/DownloadInteractor;", "(Lcom/bits/bee/bpmc/domain/repository/DistrictRepository;Lcom/bits/bee/bpmc/domain/usecase/download/DownloadInteractor;)V", "page", "", "downloadAddOn", "", "(Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "downloadAddOnD", "downloadAll", "Lkotlinx/coroutines/Job;", "downloadBp", "downloadCcType", "downloadChannel", "downloadCity", "downloadCmp", "downloadCrc", "downloadDistrict", "downloadEdc", "downloadEdcSurc", "downloadItem", "downloadItemAddOn", "downloadItemBranch", "downloadItemGroup", "downloadItemSaleTax", "downloadItemVariant", "downloadOperator", "downloadPmtd", "downloadPrice", "downloadPriceLvl", "downloadPromo", "downloadProvince", "downloadRegency", "downloadSelection", "downloadSelectionD", "downloadTax", "downloadUnit", "downloadVariant", "UIEvent", "app_debug"})
public final class DownloadViewModel extends com.bits.bee.bpmc.presentation.base.BaseViewModel<com.bits.bee.bpmc.presentation.ui.download.DownloadState, com.bits.bee.bpmc.presentation.ui.download.DownloadViewModel.UIEvent> {
    private final com.bits.bee.bpmc.domain.repository.DistrictRepository districtRepository = null;
    private final com.bits.bee.bpmc.domain.usecase.download.DownloadInteractor di = null;
    private int page = 1;
    
    @javax.inject.Inject()
    public DownloadViewModel(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.repository.DistrictRepository districtRepository, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.usecase.download.DownloadInteractor di) {
        super();
    }
    
    private final kotlinx.coroutines.Job downloadAll() {
        return null;
    }
    
    private final java.lang.Object downloadItemGroup(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadChannel(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadBp(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadPriceLvl(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadItem(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadCity(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadOperator(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadProvince(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadCrc(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadRegency(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadCmp(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadDistrict(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadPrice(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadItemBranch(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadItemSaleTax(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadTax(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadEdc(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadEdcSurc(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadCcType(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadPmtd(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadUnit(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadAddOn(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadSelection(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadSelectionD(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadAddOnD(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadItemAddOn(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadVariant(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadItemVariant(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    private final java.lang.Object downloadPromo(kotlin.coroutines.Continuation<? super kotlin.Unit> continuation) {
        return null;
    }
    
    @kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000\f\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0002\b6\u0018\u00002\u00020\u0001B\u0007\b\u0004\u00a2\u0006\u0002\u0010\u0002\u00a8\u0006\u0003"}, d2 = {"Lcom/bits/bee/bpmc/presentation/ui/download/DownloadViewModel$UIEvent;", "", "()V", "app_debug"})
    public static abstract class UIEvent {
        
        private UIEvent() {
            super();
        }
    }
}

File: app/src/main/java/com/bits/bee/bpmc/presentation/ui/download/DownloadViewModel.kt
package com.bits.bee.bpmc.presentation.ui.download

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bits.bee.bpmc.domain.mapper.DistrictDataMapper
import com.bits.bee.bpmc.domain.repository.DistrictRepository
import com.bits.bee.bpmc.domain.repository.ItemGroupRepository
import com.bits.bee.bpmc.domain.repository.ProvinceRepository
import com.bits.bee.bpmc.domain.repository.RegencyRepository
import com.bits.bee.bpmc.domain.usecase.download.DownloadInteractor
import com.bits.bee.bpmc.utils.Resource
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Created by aldi on 29/03/22.
 */

@HiltViewModel
class DownloadViewModel @Inject constructor (
    private val itemGroupRepository: ItemGroupRepository,
    private val provinceRepository: ProvinceRepository,
    private val regencyRepository: RegencyRepository,
    private val districtRepository: DistrictRepository,
    private val di: DownloadInteractor,
) : ViewModel() {

    private var page : Int = 1

    private val _state = MutableStateFlow(DownloadState())
    var state = _state.asStateFlow()

    init {
        downloadAll()
    }

    private fun downloadAll() = viewModelScope.launch {
        downloadItemGroup()
    }

    private fun downloadItemGroup() = viewModelScope.launch {
        di.getLatestItemGroupUseCase(page).collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Item Group")
                    }
                }
                Resource.Status.SUCCESS -> {
                    it.data?.let { data ->
                        _state.update {
                            it.copy(status = "Finish Downloading Item Group")
                        }

                        downloadPriceLvl()
                    }
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadChannel() = viewModelScope.launch {
        di.getLatestChannelUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Channel")
                    }

                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Channel")
                    }
                    downloadBp()
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadBp() = viewModelScope.launch {
        di.getLatestBpUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {

                    _state.update {
                        it.copy(status = "Downloading Bp")
                    }
                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Bp")
                    }

                    downloadItem()
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadPriceLvl() = viewModelScope.launch {
        di.getLatestPriceLvlUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Price Lvl")
                    }
                }
                Resource.Status.SUCCESS -> {

                    downloadChannel()
                    _state.update {
                        it.copy(status = "Finish Downloading Price Lvl")
                    }
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadItem() = viewModelScope.launch {
        di.getLatestItemUseCase(page).collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Item")
                    }
                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Item")
                    }
                    downloadPrice()
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadCity() = viewModelScope.launch {
        di.getLatestCityUseCase(page).collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading City")
                    }
                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading City")
                    }
                    downloadItemSaleTax()
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadOperator() = viewModelScope.launch {
        di.getLatestOperatorUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {

                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Operator")
                    }
                }
                Resource.Status.ERROR -> {
                }
            }
        }
    }

//    private fun downloadProvince(): Job = viewModelScope.launch {
//        downloadInteractor.getLatestProvinceUseCase.invoke(page).collect {
//            when(it.status){
//                Resource.Status.LOADING -> {
//                    _state.update {
//                        it.copy(
//                            status = "Download Province"
//                        )
//                    }
//                }
//                Resource.Status.SUCCESS ->{
//                    it.data?.let { data ->
//                        provinceRepository.insertBulkProvince(
//                            data.data.data.map { ProvinceDataMapper.fromResponseToData(it) }
//                        )
//                        page++
//                        if (data.data.page <= data.data.total_page){
//                            downloadProvince()
//                        }else{
//                            _state.update {
//                                it.copy(
//                                    status = "Finish Download Province"
//                                )
//                            }
//                            page = 1
//                            downloadRegency()
//                        }
//                    }
//                }
//                Resource.Status.ERROR ->{
//
//                }
//            }
//        }
//    }

    private fun downloadCrc() : Job = viewModelScope.launch {
        di.getLatestCrcUseCase().collect {
            when (it.status) {
                Resource.Status.LOADING -> {

                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Crc")
                    }
                    downloadPrice()
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }


//    private fun downloadRegency(): Job = viewModelScope.launch {
//        downloadInteractor.getLatestRegencyUseCase.invoke(page).collect {
//            when(it.status){
//                Resource.Status.LOADING ->{
//                    _state.update {
//                        it.copy(
//                            status = "Download Regency"
//                        )
//                    }
//                }
//                Resource.Status.SUCCESS ->{
//                    it.data?.let { data ->
//                        regencyRepository.insertBulkRegency(
//                            data.data.data.map { RegencyDataMapper.fromResponseToData(it) }
//                        )
//                        page++
//                        if (data.data.page <= data.data.total_page){
//                            downloadRegency()
//                        }else{
//                            _state.update {
//                                it.copy(
//                                    status = "Finish Download Regency"
//                                )
//                            }
//                            page = 1
//                            downloadDistrict()
//                        }
//                    }
//                }
//                Resource.Status.ERROR ->{
//
//
//                }
//            }
//        }
//    }

    private fun downloadCmp() = viewModelScope.launch {
        di.getLatestCmpUseCase().collect {
            when (it.status) {
                Resource.Status.LOADING -> {

                }
                Resource.Status.SUCCESS -> {
//                    downloadProvince()
                    _state.update {
                        it.copy(status = "Finish Downloading Cmp")
                    }
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadDistrict(): Job = viewModelScope.launch {
        di.getLatestDistrictUseCase.invoke(page).collect {
            when(it.status){
                Resource.Status.LOADING ->{
                    _state.update {
                        it.copy(
                            status = "Download District"
                        )
                    }
                }
                Resource.Status.SUCCESS ->{
                    it.data?.let { data ->
                        districtRepository.insertBulkDistrict(
                            data.data.data.map { DistrictDataMapper.fromNetworkToDb(it) }
                        )
                        page++
                        if (data.data.page <= data.data.total_page){
                            downloadDistrict()
                        }else{
                            downloadPrice()
                            _state.update {
                                it.copy(
                                    status = "Finish Download District"
                                )
                            }
                        }
                    }

                }
                Resource.Status.ERROR -> {
                    downloadPrice()
                }
            }
        }
    }

    private fun downloadPrice() = viewModelScope.launch {
        di.getLatestPriceUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Price")
                    }
                }
                Resource.Status.SUCCESS -> {
                    downloadItemBranch()
                    _state.update {
                        it.copy(status = "Finish Downloading Price")
                    }
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadItemBranch() = viewModelScope.launch {
        di.getLatestItemBranchUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Item Branch")
                    }
                }
                Resource.Status.SUCCESS -> {
                    downloadCity()
                    _state.update {
                        it.copy(status = "Finish Downloading Item Branch")
                    }
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadItemSaleTax() = viewModelScope.launch {
        di.getLatestItemSaleTaxUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Item Sale Tax")
                    }
                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Item Sale Tax")
                    }
                    downloadTax()
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadTax() = viewModelScope.launch {
        di.getLatestTaxUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Tax")
                    }
                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Tax")
                    }
                    downloadEdc()
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadEdc() = viewModelScope.launch {
        di.getLatestEdcUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Edc")
                    }
                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Edc")
                    }
                    downloadEdcSurc()
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadEdcSurc() = viewModelScope.launch {
        di.getLatestEdcSurcUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Edc Surc")
                    }
                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Edc Surc")
                    }
                    downloadCcType()
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadCcType() = viewModelScope.launch {
        di.getLatestCcTypeUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Cc Type")
                    }
                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Cc Type")
                    }
                    downloadPmtd()
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadPmtd() = viewModelScope.launch {
        di.getLatestPmtdUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Pmtd")
                    }
                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Pmtd")
                    }
                    downloadUnit()
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    private fun downloadUnit() = viewModelScope.launch {
        di.getLatestUnitUseCase().collect {
            when(it.status){
                Resource.Status.LOADING -> {
                    _state.update {
                        it.copy(status = "Downloading Unit")
                    }
                }
                Resource.Status.SUCCESS -> {
                    _state.update {
                        it.copy(status = "Finish Downloading Unit")
                    }
                }
                Resource.Status.ERROR -> {

                }
            }
        }
    }

    suspend fun awaitAll(vararg blocks: suspend () -> Unit) = viewModelScope.launch {
        blocks.forEach {
            launch { it() }
        }
    }

}